src/battle.ts

```diff
@@ -994,8 +1001,11 @@ interface ServerPokemon extends PokemonDetails, PokemonHealth {
 	pokeball: string;
 }
 
+type SceneProvider = (b: Battle) => BattleSceneStub;
+type SideProvider = (b: Battle, n: number) => Side;
+
 class Battle {
-	scene: BattleScene | BattleSceneStub;
+	scene: BattleSceneStub;
```

Comment:
Actually do we even need a stub now that we have optional chaining and can just replace `scene.` with `scene?.`?

I guess having the stub is nice for other people using it, but maybe it can be an interface?

Reply:
@scheibo accidentally edited my post and wrote

> Actually do we even need a stub now that we have optional chaining and can just replace `scene.` with `scene?.`?

Perhaps not, but changing that is large in scope than this. I could do it in a follow up if thats the pattern you wanted to switch to

> I guess having the stub is nice for other people using it, but maybe it can be an interface?

I thought about making it an interface, and that may be the way to go if we switch to an interface (and have all of the methods options), but as it stands its basically an abstract base class type of pattern that means people only need to override the methods they want to

Reply:
> Actually do we even need a stub now that we have optional chaining and can just replace `scene.` with `scene?.`?

Perhaps not, but changing that is large in scope than this. I could do it in a follow up if thats the pattern you wanted to switch to

> I guess having the stub is nice for other people using it, but maybe it can be an interface?

I thought about making it an interface, and that may be the way to go if we switch to an interface (and have all of the methods options), but as it stands its basically an abstract base class type of pattern that means people only need to override the methods they want to